commit 8f480bd2664e836beb0d82583d9ee0cfe2fadb9c
Author: Xuefer <xuefer@gmail.com>
Date:   Wed Mar 18 17:07:23 2015 +0800

    fix standby/stopInput race condition
    
    standby() is currently likely be called after stopInput()
    crashing the driver (happens on Tegra 4, Xiaomi MI 3 TD, pisces)
    
    fixed by calling standby() before condition broadcast
    
    Change-Id: Ib63aad6cda01baf0ad1e2fab59c523be1f990d42

diff --git a/services/audioflinger/Threads.cpp b/services/audioflinger/Threads.cpp
index 3192cd8..82a6ba3 100644
--- a/services/audioflinger/Threads.cpp
+++ b/services/audioflinger/Threads.cpp
@@ -5584,10 +5584,9 @@ reacquire_wakelock:
                 break;
             }
 
-            // if no active track(s), then standby and release wakelock
+            // if no active track(s), then standby (down below) and release wakelock (here)
             size_t size = mActiveTracks.size();
             if (size == 0) {
-                standbyIfNotAlreadyInStandby();
                 // exitPending() can't become true here
                 releaseWakeLock_l();
                 ALOGV("RecordThread: loop stopping");
@@ -5663,6 +5662,10 @@ reacquire_wakelock:
                     fastTrack = activeTrack;
                 }
             }
+            // standby must be done before mStartStopCond broadcast
+            if (activeTracks.size() == 0) {
+                standbyIfNotAlreadyInStandby();
+            }
             if (doBroadcast) {
                 mStartStopCond.broadcast();
             }
